### PR TITLE
Fix libcudf memory errors

### PR DESCRIPTION
**Pull request by karthikeyann**
Fixes https://github.com/rapidsai/cudf/issues/8883

All memory errors caught in libcudf unit tests are fixed in this PR.
These unit tests are checked with `cuda-memcheck` before and after the fix.

The following tests FAILED:
- [x]          2 - SCALAR_TEST (Failed) (Fixed)
- [x]         32 - COPYING_TEST (Failed) (Fixed)
- [x]         39 - MERGE_TEST (Failed) (Fixed)
- [x]         46 - FACTORIES_TEST (Failed) (Fixed by scalar test fix)

### CODE REVIEW

**Comment by karthikeyann** [comment]
@nvdbaranec  I am debugging an out of memory access failure.
In the following 2 lines in unit tests,
https://github.com/rapidsai/cudf/blob/7de2dc811f956461cfe68f1d518d46a841b099c5/cpp/tests/copying/concatenate_tests.cu#L485

https://github.com/rapidsai/cudf/blob/7de2dc811f956461cfe68f1d518d46a841b099c5/cpp/tests/copying/concatenate_tests.cu#L516

Should the chars column size be `total_chars_size` ?

File: cpp/tests/utilities/column_utilities.cu [hunk 1]
```diff
@@ -118,9 +118,12 @@ std::unique_ptr<column> generate_child_row_indices(lists_column_view const& c,
     0,
     [row_indices = row_indices.begin<size_type>(),
      validity    = c.null_mask(),
+     offsets     = c.offsets().begin<offset_type>(),
      offset      = c.offset()] __device__(int index) {
       auto const true_index = row_indices[index] + offset;
-      return !validity || cudf::bit_is_set(validity, true_index) ? 1 : 0;
+      return !validity || cudf::bit_is_set(validity, true_index)
+               ? (offsets[true_index + 1] - offsets[true_index]) != 0
+               : 0;
```
Comment (nvdbaranec) [inline cpp/tests/utilities/column_utilities.cu:126]: I'm not sure I follow the logic here.  This change would seem to add nulls anywhere there is an empty row, which is not correct.

**Review by nvdbaranec — DISMISSED** [review]

**Comment by nvdbaranec** [comment]
> 
> 
> @nvdbaranec I am debugging an out of memory access failure.
> In the following 2 lines in unit tests,
> 
> https://github.com/rapidsai/cudf/blob/7de2dc811f956461cfe68f1d518d46a841b099c5/cpp/tests/copying/concatenate_tests.cu#L485
> 
> https://github.com/rapidsai/cudf/blob/7de2dc811f956461cfe68f1d518d46a841b099c5/cpp/tests/copying/concatenate_tests.cu#L516
> 
> Should the chars column size be `total_chars_size` ?

Correct, that looks like the right fix.

**Comment by karthikeyann** [comment]
rerun tests

File: cpp/tests/utilities/column_utilities.cu [hunk 2]
```diff
@@ -114,13 +114,16 @@ std::unique_ptr<column> generate_child_row_indices(lists_column_view const& c,
   //
   // result = [6, 1, 11, 1, 1]
   //
-  auto validity_iter = cudf::detail::make_counting_transform_iterator(
+  auto is_output_row = cudf::detail::make_counting_transform_iterator(
```
Comment (nvdbaranec) [inline cpp/tests/utilities/column_utilities.cu:117]: What I was suggesting that this iterator and the one above (`row_size_iter`) are the same and should be combined.

**Review by cwharris — APPROVED** [review]

**Review by davidwendt — APPROVED** [review]

**Review by mythrocks — APPROVED** [review]
LGTM

File: cpp/tests/utilities/column_utilities.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/tests/utilities/column_utilities.cu]: reused `row_size_iter` and use a stencil predicate in `scatter_if`.

**Comment by karthikeyann** [comment]
rerun tests

**Comment by karthikeyann** [comment]
rerun tests

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**